pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -730,13 +729,47 @@ public void createConsumingSegment(String realtimeTableName, List<Integer> nonCo
           int nextSeqNum = segmentName.getSequenceNumber() + 1;
           List<String> instances = partitionAssignment.getListField(Integer.toString(curPartition));
           LOGGER.info("Creating CONSUMING segment for {} partition {} with seq {}", realtimeTableName, curPartition, nextSeqNum);
+          // To begin with, set startOffset to the oldest availble offset in kafka. Fix it to be the one we want,
+          // depending on what the prev segment had.
           long startOffset = getKafkaPartitionOffset(kafkaStreamMetadata, "smallest", curPartition);
           LOGGER.info("Found kafka offset {} for table {} for partition {}", startOffset, realtimeTableName, curPartition);
+
           final LLCRealtimeSegmentZKMetadata oldSegMetadata = getRealtimeSegmentZKMetadata(realtimeTableName,
               segmentName.getSegmentName());
-          if (startOffset < oldSegMetadata.getEndOffset()) {
-            startOffset = oldSegMetadata.getEndOffset();
-            LOGGER.info("Choosing newer kafka offset {} for table {} for partition {}", startOffset, realtimeTableName, curPartition);
+          CommonConstants.Segment.Realtime.Status status = oldSegMetadata.getStatus();
+          final long prevSegStartOffset = oldSegMetadata.getStartOffset();  // Offset at which the prev segment intended to start consuming
+          if (status.equals(CommonConstants.Segment.Realtime.Status.IN_PROGRESS)) {
```

Comment:
This loop is really too convoluted. It's trying to iterate in order through the segments of all partitions and tries to:

* Fix missing partitions
* Find out the max valid offset
* Fix missing segments

This would be significantly more readable if this was broken up into smaller, individually readable parts.